src/Search/TreeTimesyncBeamSearch/TreeTimesyncBeamSearch.cc

```diff
@@ -0,0 +1,696 @@
+    featureProcessingTime_.stop();
+}
+
+void TreeTimesyncBeamSearch::putFeatures(Nn::DataView const& features, size_t nTimesteps) {
+    featureProcessingTime_.start();
+    labelScorer_->addInputs(features, nTimesteps);
+    featureProcessingTime_.stop();
+}
+
+Core::Ref<const Traceback> TreeTimesyncBeamSearch::getCurrentBestTraceback() const {
+    return getBestHypothesis().trace->performTraceback();
+}
+
+Core::Ref<const LatticeAdaptor> TreeTimesyncBeamSearch::getCurrentBestWordLattice() const {
+    auto&        bestHypothesis = getBestHypothesis();
+    LatticeTrace endTrace(bestHypothesis.trace, 0, bestHypothesis.trace->time + 1, bestHypothesis.trace->score, {});
+
+    for (size_t hypIdx = 1ul; hypIdx < beam_.size(); ++hypIdx) {
+        auto& hyp          = beam_[hypIdx];
+        auto  siblingTrace = Core::ref(new LatticeTrace(hyp.trace, 0, hyp.trace->time, hyp.trace->score, {}));
+        endTrace.appendSiblingToChain(siblingTrace);
+    }
+
+    return endTrace.buildWordLattice(lexicon_);
+}
+
+bool TreeTimesyncBeamSearch::decodeStep() {
+    if (finishedSegment_) {
+        return false;
+    }
+
+    /*
+     * Collect all possible within-word extensions for all hypotheses in the beam.
+     * Also create scoring requests for the label scorer.
+     * Each extension candidate makes up a request.
+     */
+    extensions_.clear();
+    requests_.clear();
+
+    for (size_t hypIndex = 0ul; hypIndex < beam_.size(); ++hypIndex) {
+        auto& hyp = beam_[hypIndex];
+
+        // Iterate over the successors of this hypothesis' current state in the tree
+        for (const auto& successorState : stateSuccessorLookup_[hyp.currentState]) {
+            Nn::LabelIndex tokenIdx = network_->structure.state(successorState).stateDesc.acousticModel;
+            // If we want to force blank between repeated labels across words, a new word should not start with the same token as the previous word ended (except for blank itself)
+            // If we don't force blank and we have a repeated label across words, we need to make sure to have label-to-Label as transition type
+            if (!(
+                        forceBlankAcrossWords_ &&
+                        hyp.currentState == network_->rootState &&
+                        tokenIdx == hyp.currentToken &&
+                        (!useBlank_ || tokenIdx != blankLabelIndex_))) {
+                auto transitionType = inferTransitionType(hyp.currentToken, tokenIdx, hyp.currentState == network_->rootState);
+                extensions_.push_back(
+                        {tokenIdx,
+                         nullptr,
+                         successorState,
+                         hyp.lmHistory,
+                         hyp.score,
+                         0.0,
+                         0,
+                         transitionType,
+                         hypIndex});
+                requests_.push_back({beam_[hypIndex].scoringContext, tokenIdx, transitionType});
+            }
```

Comment:
```suggestion
            if (forceBlankAcrossWords_ and
                hyp.currentState == network_->rootState and
                tokenIdx == hyp.currentToken and
                (not useBlank_ or tokenIdx != blankLabelIndex_)) {
                continue;
            }
            auto transitionType = inferTransitionType(hyp.currentToken, tokenIdx, hyp.currentState == network_->rootState);
            extensions_.push_back(
                    {tokenIdx,
                     nullptr,
                     successorState,
                     hyp.lmHistory,
                     hyp.score,
                     0.0,
                     0,
                     transitionType,
                     hypIndex});
            requests_.push_back({beam_[hypIndex].scoringContext, tokenIdx, transitionType});

```
The formatting is a bit weird and negating the entire expression makes it difficult to follow. It would be better to formulate it as a guard clause.